dtfs-central-api/src/services/repositories/utilisation-reports-repo/utilisation-reports-repo.ts

```diff
@@ -0,0 +1,46 @@
+import { ObjectId, OptionalId } from 'mongodb';
+import sortBy from 'lodash/sortBy';
+import db from '../../../drivers/db-client';
+import { UTILISATION_REPORT_RECONCILIATION_STATUS, DB_COLLECTIONS } from '../../../constants';
+import { AzureFileInfo } from '../../../types/azure-file-info';
+import { UtilisationReport } from '../../../types/db-models/utilisation-reports';
+import { PortalSessionUser } from '../../../types/portal/portal-session-user';
+
+export const saveUtilisationReportDetails = async (month: number, year: number, azureFileInfo: AzureFileInfo, uploadedByUser: PortalSessionUser) => {
+  const utilisationReportInfo: OptionalId<UtilisationReport> = {
+    bank: {
+      id: uploadedByUser.bank.id,
+      name: uploadedByUser.bank.name,
+    },
+    month: Number(month),
+    year: Number(year),
+    dateUploaded: new Date(),
+    azureFileInfo,
+    status: UTILISATION_REPORT_RECONCILIATION_STATUS.PENDING_RECONCILIATION,
+    uploadedBy: {
+      id: uploadedByUser._id,
+      firstname: uploadedByUser.firstname,
+      surname: uploadedByUser.surname,
+    },
+  };
+
+  const utilisationReportDetailsCollection = await db.getCollection(DB_COLLECTIONS.UTILISATION_REPORTS);
+  const savedDetails = await utilisationReportDetailsCollection.insertOne(utilisationReportInfo);
+  return { reportId: savedDetails.insertedId.toString(), dateUploaded: utilisationReportInfo.dateUploaded };
+};
+
+export const getUtilisationReportDetailsByBankIdMonthAndYear = async (bankId: string, month: number, year: number): Promise<UtilisationReport | null> => {
```

Comment:
We need to explicitly define the return types for these functions as the `getCollection` returns something like `Promise<Collection<WithoutId<DbCollection>>>`, but the call of `collection.find(...).toArray()` (for example) then returns `WithId<WithoutId<DbCollection>>[]` which is, of course, the same as `DbCollection[]`. In order for typescript to recognise this, however, I need to explicitly provide the return type. I have made this change to almost all the functions that query the database and return something, so any suggestions for a better way to do this will be applied to all those locations

Reply:
I think it only happens because we have been "smart" in including the `_id` field in the `db-models` explicitly via `WithId`, so we then have to remove the `_id` for the collection type definition. But then, the `collection` adds it back once a method on the `collection` is called which returns something. We would get better inferred types for things like `getCollection().method()` if we removed the `WithId` from the `db-models` type, but we would then have to add `WithId` wherever the `db-models` types are used outside of the `collection` definitions, so I think your type definitions in `getCollection` are fine as they are